consensus/wal_iter_test.go

```diff
@@ -0,0 +1,188 @@
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1002, Round: 0}},
+			},
+			input: []interface{}{
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 0, Round: 0}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1000, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1000, Round: 0}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1001, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1000, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1000, Round: 1}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1001, Round: 1}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 1}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1001, Round: 2}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 1}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1001, Round: 3}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 1}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 2}},
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1002, Round: 0}},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1002, Round: 0}},
+			},
+		},
+		{
+			input: generateMessageSequence(1000, 1002, 1000),
+			want: []Message{
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1000, Round: 1000}},
+				&BlockPartMessage{Height: 1000, Round: 1000},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1000, Round: 1000}},
+
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1001, Round: 1000}},
+				&BlockPartMessage{Height: 1001, Round: 1000},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1001, Round: 1000}},
+
+				&ProposalMessage{Proposal: &tmtypes.Proposal{Height: 1002, Round: 1000}},
+				&BlockPartMessage{Height: 1002, Round: 1000},
+				&VoteMessage{Vote: &tmtypes.Vote{Height: 1002, Round: 1000}},
+			},
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("test-case #%d", i), func(t *testing.T) {
+			it := newWalIter(&mockWALReader{ret: tc.input})
+			cnt := 0
+			for it.Next() {
+				msg := it.Value()
+				mi := msg.Msg.(msgInfo)
+				require.Equal(t, tc.want[cnt], mi.Msg)
+				cnt++
+			}
+			require.NoError(t, it.Err())
+			require.Equal(t, len(tc.want), cnt)
+		})
+	}
+}
+
+func TestWalIter_Err(t *testing.T) {
+	testCases := []struct {
+		err  error
+		want string
+	}{
+		{
+			err: io.EOF,
+		},
+		{
+			err:  errors.New("reader error"),
+			want: "reader error",
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("test-case #%d", i), func(t *testing.T) {
+			it := newWalIter(&mockWALReader{ret: []interface{}{
+				tc.err,
+			}})
+			require.False(t, it.Next())
+			requireError(t, tc.want, it.Err())
+		})
+	}
+}
+
+func requireError(t *testing.T, want string, err error) {
+	if err != nil {
```

Comment:
maybe `require.NotNil(t,err,want)` ?

Reply:
we need a bit complex assertions
1. if `error` is not nil then the error must match on expected string
2. otherwise, `want` must be empty string